### PR TITLE
Add logging

### PR DESCRIPTION
**Pull request by Sammers21**
implementation of https://github.com/vert-x3/vertx-mqtt-client/issues/5 enhancement

### CODE REVIEW

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 1]
```diff
@@ -116,6 +118,10 @@ public MqttClient connect(Handler<AsyncResult<MqttConnAckMessage>> connectHandle
         }
       } else {
 
+        if (options.getLogActivity()) {
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: the getLogActivity() method is for low level log inside the net client. it's used just for adding the "logging" in the pipeline as you have alredy done. The log I'd like to have is just enabled with log4j (trace, debug, info and so on). So use log.debug without the "if" on the getLogActivity().

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 2]
```diff
@@ -216,7 +230,12 @@ public MqttClient publish(String topic, Buffer payload, MqttQoS qosLevel, boolea
 
     io.netty.handler.codec.mqtt.MqttMessage publish = MqttMessageFactory.newMessage(fixedHeader, variableHeader, buf);
 
-    this.write(publish);
+    this.write(publish).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info(String.format("Clint have just sent PUBLISH packet with topic=%s,qos=%d,isDup=%s,isRetain=%s",
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: use log.debug and remove the "if". Remember that log.XXX methods have a syntax which uses {} as placeholder no need for String.format. Finally, we need a toString() method in the messages we have to print. For native MQTT netty message we should already have them. I'm going to add toString() for message classes coming from MQTT server you can do the same for messages classes we have defined here.

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 3]
```diff
@@ -302,7 +321,11 @@ public MqttClient subscribe(Map<String, Integer> topics, Handler<AsyncResult<Int
 
     io.netty.handler.codec.mqtt.MqttMessage subscribe = MqttMessageFactory.newMessage(fixedHeader, variableHeader, payload);
 
-    this.write(subscribe);
+    this.write(subscribe).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent SUBSCRIBE packet with " + topics.size() + " pairs of topic and QoS level");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 4]
```diff
@@ -339,7 +362,11 @@ public MqttClient unsubscribe(String topic, Handler<AsyncResult<Integer>> unsubs
 
     io.netty.handler.codec.mqtt.MqttMessage unsubscribe = MqttMessageFactory.newMessage(fixedHeader, variableHeader, payload);
 
-    this.write(unsubscribe);
+    this.write(unsubscribe).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent UNSUBSCRIBE packet");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 5]
```diff
@@ -376,7 +403,11 @@ public MqttClient ping() {
 
     io.netty.handler.codec.mqtt.MqttMessage pingreq = MqttMessageFactory.newMessage(fixedHeader, null, null);
 
-    this.write(pingreq);
+    this.write(pingreq).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent PINGREQ packet");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 6]
```diff
@@ -396,7 +427,11 @@ void publishAcknowledge(int publishMessageId) {
 
     io.netty.handler.codec.mqtt.MqttMessage puback = MqttMessageFactory.newMessage(fixedHeader, variableHeader, null);
 
-    this.write(puback);
+    this.write(puback).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent PUBACK packet with messageId=" + publishMessageId);
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 7]
```diff
@@ -414,7 +449,11 @@ void publishReceived(int publishMessageId) {
 
     io.netty.handler.codec.mqtt.MqttMessage pubrec = MqttMessageFactory.newMessage(fixedHeader, variableHeader, null);
 
-    this.write(pubrec);
+    this.write(pubrec).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent PUBREC packet with messageId=" + publishMessageId);
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 8]
```diff
@@ -432,7 +471,11 @@ void publishComplete(int publishMessageId) {
 
     io.netty.handler.codec.mqtt.MqttMessage pubcomp = MqttMessageFactory.newMessage(fixedHeader, variableHeader, null);
 
-    this.write(pubcomp);
+    this.write(pubcomp).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent PUBCOMP packet with messageId=" + publishMessageId);
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 9]
```diff
@@ -450,7 +493,11 @@ void publishRelease(int publishMessageId) {
 
     io.netty.handler.codec.mqtt.MqttMessage pubrel = MqttMessageFactory.newMessage(fixedHeader, variableHeader, null);
 
-    this.write(pubrel);
+    this.write(pubrel).addListener((ChannelFutureListener) channelFuture -> {
+      if (options.getLogActivity()) {
+        log.info("Clint have just sent PUREL packet with messageId=" + publishMessageId);
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 10]
```diff
@@ -555,11 +606,10 @@ private int nextMessageId() {
     return this.messageIdCounter;
   }
 
-  public MqttClientImpl write(io.netty.handler.codec.mqtt.MqttMessage mqttMessage) {
+  public ChannelFuture write(io.netty.handler.codec.mqtt.MqttMessage mqttMessage) {
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: why this change ?

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 11]
```diff
@@ -569,6 +619,10 @@ public MqttClientImpl write(io.netty.handler.codec.mqtt.MqttMessage mqttMessage)
   void handlePingresp() {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("PINGRESP received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 12]
```diff
@@ -583,6 +637,10 @@ void handlePingresp() {
   void handleUnsuback(int unsubackMessageId) {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("UNSUBACK received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 13]
```diff
@@ -597,6 +655,10 @@ void handleUnsuback(int unsubackMessageId) {
   void handlePuback(int pubackMessageId) {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("PUBACK received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 14]
```diff
@@ -611,6 +673,10 @@ void handlePuback(int pubackMessageId) {
   void handlePubcomp(int pubcompMessageId) {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("PUBCOMP received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 15]
```diff
@@ -625,6 +691,10 @@ void handlePubcomp(int pubcompMessageId) {
   void handlePubrec(int pubrecMessageId) {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("PUBREC received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 16]
```diff
@@ -637,6 +707,10 @@ void handlePubrec(int pubrecMessageId) {
   void handleSuback(MqttSubAckMessage msg) {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("SUBACK with qos levels for " + msg.grantedQoSLevels().size() + " topics received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 17]
```diff
@@ -651,7 +725,9 @@ void handleSuback(MqttSubAckMessage msg) {
   void handlePublish(MqttPublishMessage msg) {
 
     synchronized (this.connection) {
-
+      if (options.getLogActivity()) {
+        log.info("PUBLISH with qos level " + msg.qosLevel().value() + " received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 18]
```diff
@@ -684,6 +760,9 @@ void handlePublish(MqttPublishMessage msg) {
   void handlePubrel(int pubrelMessageId) {
 
     synchronized (this.connection) {
+      if (options.getLogActivity()) {
+        log.info("PUBREL received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 19]
```diff
@@ -699,7 +778,9 @@ void handlePubrel(int pubrelMessageId) {
   void handleConnack(MqttConnAckMessage msg) {
 
     synchronized (this.connection) {
-
+      if (options.getLogActivity()) {
+        log.info("CONNACK received");
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: same as previous comment

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 1 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Does it means that i should use log4j instead of io.vertx.core.logging.Logger ? 

Also, i should have a additional flag for high-level logging?

Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Yes I wouldn't use the logger class from Vert.x for that sl4j and relying on a log4j.properties file (or command line parameters for that) when you want to specify the log level.

**Comment by ppatierno** [comment]
@Sammers21 reviewing some code even in the MQTT server I think that we can relying on the toString() provided by the MQTT messages classes from Netty. It means no need to implement toString() for our Vert.x messages classes. The safeObject() method is where we can trace the incoming MQTT messages and the write() where we can trace out messages. In both cases we have an io.netty.handler.codec.mqtt.MqttMessage instance with toString() already implemented. This is about tracing MQTT messages, feel free to find some other points where having a log.error or log.warn or log.info could make sense at MQTT client level. Always using sl4j.

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 10 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: this was changed because i want to use addlistener method for logging :-)

Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: or we don't need to be sure that server receive message before we log a even of sending?

Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: not for now ;)

**Comment by Sammers21** [comment]
@ppatierno , I have just updated the repo. 

* slf4j + log4j2  for logging.
* logging of outcoming messages via an ChannelOutboundHandler .
* Incoming messages loggin in safeObject method.

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 20]
```diff
@@ -106,15 +104,17 @@ public MqttClient connect() {
   @Override
   public MqttClient connect(Handler<AsyncResult<MqttConnAckMessage>> connectHandler) {
 
+    log.debug("trying to connect with {}:{}", options.getHost(), options.getPort());
     this.doConnect(options.getPort(), options.getHost(), options.getHost(), done -> {
 
       // the TCP connection fails
       if (done.failed()) {
-
+        log.warn("Can't connect to {}:{}", options.getHost(), options.getPort());
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: It should be an error not a warn. Wdyt ?

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 21]
```diff
@@ -699,16 +706,16 @@ void handlePubrel(int pubrelMessageId) {
...
     synchronized (this.connection) {
-
       if (this.connectHandler != null) {
 
         if (msg.code() == MqttConnectReturnCode.CONNECTION_ACCEPTED) {
           this.connectHandler.handle(Future.succeededFuture(msg));
         } else {
-          this.connectHandler.handle(Future.failedFuture(new MqttConnectionException(msg.code())));
+          MqttConnectionException exception = new MqttConnectionException(msg.code());
+          log.warn("{}", exception);
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Error not warn. When you need to pass the exception object you don't need placeholder. At least write an error message without placeholder.

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 22]
```diff
@@ -504,6 +511,7 @@ protected Object safeObject(Object msg, ByteBufAllocator allocator) {
       DecoderResult result = mqttMessage.decoderResult();
       if (result.isSuccess() && result.isFinished()) {
 
+        log.info("Incoming packet {}", msg);
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Use debug not info. In general info is the default level but we want message output only at debug level.

Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: sure, changed to debug

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 21 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: done

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 20 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: You are right. i think that failed attempt to connect is not a warning, it is a significant error

Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Done

File: src/main/resources/log4j2.xml [hunk 23]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ppatierno) [inline src/main/resources/log4j2.xml]: we don't need to put the log4j configuration in the "main" source code. If you want to add the file it's worth to do that in the test/resources because providing the log4j configuration file isn't something that the MQTT client has to provide but the user application. Finally, using the log4j.properties format could be better ... more readable that an XML :-)

File: src/main/java/io/vertx/mqtt/MqttOutcomingMessagesLogger.java [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.mqtt.MqttMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Outcoming messages logger
+ */
+public class MqttOutcomingMessagesLogger extends ChannelOutboundHandlerAdapter {
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttOutcomingMessagesLogger.java]: what is the advantage to use this approach with a new handler class just for logging ? isn't it simpler to have log just before the write channel inside the write method ? Just asking ...

Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/MqttOutcomingMessagesLogger.java]: Oh.... for sure... logging inside the write method is much simpler and better approach. I will do so

File: src/main/resources/log4j2.xml (hunk 23 above)
Comment (Sammers21) [inline src/main/resources/log4j2.xml]: Okay, I have just moved it to test/resources directory.

Comment (Sammers21) [inline src/main/resources/log4j2.xml]: Should I reflect it in documentation? I mean the way a user can enable logs in his application?

File: src/main/java/io/vertx/mqtt/MqttClientOptions.java [hunk 25]
```diff
@@ -4,6 +4,8 @@
 import io.vertx.codegen.annotations.DataObject;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.net.NetClientOptions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/MqttClientOptions.java]: we don't have logging here right ? Are these two unused imports ?

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 26]
```diff
@@ -106,15 +104,17 @@ public MqttClient connect() {
   @Override
   public MqttClient connect(Handler<AsyncResult<MqttConnAckMessage>> connectHandler) {
 
+    log.debug("trying to connect with {}:{}", options.getHost(), options.getPort());
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Uppercase "T" as all the other logging messages which start with uppercase first character

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 27]
```diff
@@ -106,15 +104,17 @@ public MqttClient connect() {
...
     this.doConnect(options.getPort(), options.getHost(), options.getHost(), done -> {
 
       // the TCP connection fails
       if (done.failed()) {
-
+        log.error("Can't connect to {}:{}", options.getHost(), options.getPort());
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: add the done.cause() as latest parameter of the log.error so that the exception information are printed

Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: done

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 26 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Done

File: src/main/java/io/vertx/mqtt/MqttClientOptions.java (hunk 25 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/MqttClientOptions.java]: sure

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 28]
```diff
@@ -37,18 +35,16 @@
...
+import io.vertx.mqtt.*;
 import io.vertx.mqtt.messages.MqttPublishMessage;
 
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: My apologies Pavel, I think that it's better to come back to the previous usage of Vert.x logger. I noticed that other Vert.x projects do the same. Remember that {} placeholder is not supported so you need concatenation. Sorry for that. After this changes I'll merge the PR ;)
Remember to squash all the commits.

File: pom.xml [hunk 29]
```diff
@@ -82,6 +84,28 @@
...
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-api</artifactId>
+      <version>${log4j.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-core</artifactId>
+      <version>${log4j.version}</version>
+    </dependency>
```
Comment (ppatierno) [inline pom.xml]: Reverting back to Vert.x logger you can remove log4j dependencies here leaving only the slf4j with "provided" as scope. It should be enough.

File: pom.xml [hunk 30]
```diff
@@ -82,6 +84,28 @@
       <artifactId>vertx-unit</artifactId>
       <scope>test</scope>
     </dependency>
+
+    <!-- Slf4j + log4j2 -->
+    <dependency>
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-api</artifactId>
+      <version>${slf4j.version}</version>
```
Comment (ppatierno) [inline pom.xml]: adding "provided" as scope

File: pom.xml [hunk 31]
```diff
@@ -27,6 +27,8 @@
   <properties>
     <stack.version>3.4.2-SNAPSHOT</stack.version>
     <junit.version>4.12</junit.version>
+    <slf4j.version>1.7.25</slf4j.version>
```
Comment (ppatierno) [inline pom.xml]: It's better to be aligned with the Vert.x used version which is 1.7.21

**Comment by Sammers21** [comment]
squashed

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java [hunk 32]
```diff
@@ -106,15 +104,17 @@ public MqttClient connect() {
   @Override
   public MqttClient connect(Handler<AsyncResult<MqttConnAckMessage>> connectHandler) {
 
+    log.debug(String.format("Trying to connect with %s:%d", options.getHost(), options.getPort()));
     this.doConnect(options.getPort(), options.getHost(), options.getHost(), done -> {
 
       // the TCP connection fails
       if (done.failed()) {
-
+        log.error(String.format("Can't connect to %s:%d\nCause:%s", options.getHost(), options.getPort(), done.cause().toString()));
```
Comment (ppatierno) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: It seems that Vert.x logger supports passing the exception object (done.cause()). Can you check ?

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 28 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: :-)

File: pom.xml [hunk 33]
```diff
@@ -27,6 +27,8 @@
   <properties>
     <stack.version>3.4.2-SNAPSHOT</stack.version>
     <junit.version>4.12</junit.version>
+    <slf4j.version>1.7.21</slf4j.version>
+    <log4j.version>2.8.2</log4j.version>
```
Comment (ppatierno) [inline pom.xml]: We can remove this

File: src/test/resources/log4j2.properties [hunk 34]
```diff
@@ -0,0 +1,13 @@
+status = warn
```
Comment (ppatierno) [inline src/test/resources/log4j2.properties]: Sorry but we can remove this file as well :-)

File: src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java (hunk 32 above)
Comment (Sammers21) [inline src/main/java/io/vertx/mqtt/impl/MqttClientImpl.java]: Sure it does supports
image

File: src/test/resources/log4j2.properties (hunk 34 above)
Comment (Sammers21) [inline src/test/resources/log4j2.properties]: Yes

File: pom.xml (hunk 33 above)
Comment (Sammers21) [inline pom.xml]: right

**Comment by ppatierno** [comment]
I still see conflicts here. Can you fix them ?

**Comment by Sammers21** [comment]
@ppatierno , I have just fixed it
